From a7b4dbb144c8dfa788a087298a0b47183e92a107 Mon Sep 17 00:00:00 2001
From: 6t8k <58048945+6t8k@users.noreply.github.com>
Date: Sun, 5 Nov 2023 17:01:31 +0100
Subject: [PATCH] Skip `tests/test_sidebar` if GTK initialisation fails

This may happen in headless environments.

The return value of `gtk_init_check` was not examined before; according
to [1], "calling any GTK function or instantiating any GTK type after
this function returns FALSE results in undefined behavior."

The value 77 can be returned to tell Automake and Meson that the test
was skipped.[2][3][4]

Fixes #3674

[1] https://docs.gtk.org/gtk3/func.init_check.html
[2] https://www.gnu.org/software/automake/manual/automake.html#index-Exit-status-77_002c-special-interpretation
[3] https://docs.gtk.org/glib/func.test_run.html
[4] https://mesonbuild.com/Unit-tests.html#skipped-tests-and-hard-errors
---
 tests/test_sidebar.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/tests/test_sidebar.c b/tests/test_sidebar.c
index 9659dc72e7..c03d1aa649 100644
--- a/tests/test_sidebar.c
+++ b/tests/test_sidebar.c
@@ -137,12 +137,21 @@ void test_sidebar_openfiles_tree(void)
 
 int main(int argc, char **argv)
 {
-	g_test_init(&argc, &argv, NULL);
-	/* Not sure if we can really continue without DISPLAY. Fake X display perhaps?
+	/* In headless environments that do not provide a virtual display,
+	 * GTK initialisation is expected to fail and gtk_init_check()
+	 * will return FALSE. Calling any GTK function or instantiating
+	 * any GTK type afterwards results in undefined behavior.
 	 *
-	 * This test seems to work, at least.
+	 * By returning 77, we can tell test harnesses compatible with
+	 * the GNU standard approach to regard the test as skipped.
 	 */
-	gtk_init_check(&argc, &argv);
+	if (! gtk_init_check(&argc, &argv))
+	{
+		g_test_message("GTK initialization failed; skipping. Running inside a headless environment?");
+		return 77;
+	}
+
+	g_test_init(&argc, &argv, NULL);
 
 	main_init_headless();
